urlb44e9843ba_QuasimodoNZ_SWEN222_Game_Project_tgz-pJ8-Game_World_GameWorldLevelBuilderJ8/src/Game_World/GameWorld.java

```diff
@@ -0,0 +1,476 @@
+	 *
+	 * @author Danesh Abeyratne
+	 *
+	 */
+
+	// Fields:
+	private GameEngine engine;
+	private HashMap<String, Location> locations;// the level is read into this Map
+	private HashMap<Avatar, Location> avatarLocations;// Keyset = Avatars, Values = Locations
+	private ArrayList<Avatar> avatars;
+	private int number;
+	private boolean win = false;
+	private int winner;
+
+	// Constructors:
+	public GameWorld(GameEngine engine) {// default constructor
+		this.engine = engine;
+		locations = new HashMap<String, Location>();
+		avatarLocations = new HashMap<Avatar, Location>();
+		avatars = new ArrayList<Avatar>();
+	}
+
+	public GameWorld() {
+		locations = new HashMap<String, Location>();
+		avatarLocations = new HashMap<Avatar, Location>();
+		avatars = new ArrayList<Avatar>();
+	}
+
+	public GameWorld(HashMap<String, Location> locations,
+			HashMap<Avatar, Location> avatarLocations, ArrayList<Avatar> avatars) {
+		this.locations = locations;
+		this.avatarLocations = avatarLocations;
+		this.avatars = avatars;
+	}
+
+	/**
+	 * Creates a new player in this world, and returns their UID
+	 * @param String name
+	 * @return uid of Player
+	 */
+	public int newPlayer(String name) {
+		ArrayList<String> quests = new ArrayList<String>();
+		Player p = new Player(name, quests);
+		p.setNumber(number);
+		number++;
+
+		Avatar a = null;
+		for(Avatar avatar : avatars){
+			if(avatar.getName().equals(name)){
+				a = avatar;
+				String[] sprite = a.getPlayerSprite().split("-");
+				a.setPlayerSprite(sprite[0]);
+				a.setStepsize(4);
+			}
+		}
+
+		if(a == null){
+			a = new Avatar(name, this.getAvatars().size());
+			a.setPlayerID(p.getNumber());
+			a.setLocationName("Inside Home");
+			a.setWorld(this);// add avatar to the first level
+			locations.get("Inside Home").getAllObjects().add(a);
+			p.setCharacter(a);
+			avatars.add(a);
+			//System.out.println(a);
+			avatarLocations.put(a, locations.get("Inside Home"));
+		}
+		return avatars.size();
+	}
+
+	/**
+	 * Disconnects a player from the game by removing them from the map of
+	 * Avatars.
+	 * @param uid
+	 * @return boolean
+	 */
+	public boolean disconnectPlayer(String name) {
+		Avatar a = null;
+		for (Avatar av : avatars) {
+			if (av.getName().equals(name)) {
+				a = av;
+			}
+		}
+		if(a != null){
+			a.setPlayerSprite(a.getPlayerSprite() + "-transparent");
+			a.setStepsize(0);
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * Moves a player (name) in a direction, given by the int passed
+	 * @param name Player to move
+	 * @param comd Direction to move in
+	 */
+	public void move(String name, int comd) {
+		if (comd == 37 || comd == 38 || comd == 39 || comd == 40) { // comd is a
+			// movement,
+			// so move
+			// the
+			// player
+			for (Avatar a : avatars) {
+				if (a.getName().equals(name)) {
+					a.move(comd);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Adds a Relic GameObject to the Altar
+	 * @param String name
+	 */
+	public void addRelic(String name){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 160-178 into a separate method.